addon/-private/extensions/with-validation.js

```diff
@@ -0,0 +1,35 @@
+import { wrapField } from '../wrap-field';
+import { getValidationsFor } from '../validations-for';
+
+const HAS_VALIDATION = Symbol();
+
+export function hasExtension(klass) {
+  return klass[HAS_VALIDATION];
+}
+
+/**
+ * Extend the class using native classes to inject validation logic
+ */
+export function withExtension(klass) {
+  if (klass[HAS_VALIDATION]) {
+    return klass;
+  }
+
+  klass[HAS_VALIDATION] = true;
+
+  return class extends klass {
+    static get name() {
+      return klass.name;
+    }
+
+    init(...args) {
+      super.init(...args);
+
+      const validations = getValidationsFor(this.constructor);
+
+      for (let key in validations) {
+        wrapField(this.constructor, this, validations, key);
```

Comment:
I'm a bit hesitant about this, since we're adding a wrapper to each instance of the class. Definitely think this could affect perf at scale. Even though this is not done in prod, at some point it'll become problematic even in dev.

We can iterate on this, but I think ideally we would decorate the getters themselves in place so we only need to define the properties once.

Reply:
That's a good point -- would you be OK with that being work that's done after this PR lands? It's massive as is, and I'd love to land it with a passing test suite and then work on refactoring a few things that won't break the public API, but can reduce the amount of code or improve performance.

Happy to do those things before we cut a `1.0.0` too

Reply:
Yeah, totally cool to work on it after this PR, as long as it gets done before 1.0.0 I think. That would be the last chunk of really necessary work before release.